http/agent.go

```diff
@@ -308,3 +319,221 @@ func (a *Agent) PostToWriter(w io.Writer, url string, postData []byte) error {
 	}
 	return a.readResponse(resp, w)
 }
+
+// GetRequestGroup behaves like agent.SendGetRequest() but takes a group of URLs
+// and performs the requests in parallel. The number of simultaneous requests is
+// controlled by options.MaxParallel.
+func (a *Agent) GetRequestGroup(urls []string) ([]*http.Response, []error) {
+	t := throttler.New(int(a.options.MaxParallel), len(urls))
+	ret := []*http.Response{}
+	errs := []error{}
+	m := sync.Mutex{}
+	for _, u := range urls {
+		url := u
+		func(url string) {
+			//nolint: bodyclose // We don't close here as we're returning the response
+			resp, err := a.AgentImplementation.SendGetRequest(a.Client(), url)
+
+			m.Lock()
+			errs = append(errs, err)
+			//nolint: bodyclose // We don't close here as we're returning the response
+			ret = append(ret, resp)
+			m.Unlock()
+
+			t.Done(err)
+		}(url)
+		t.Throttle()
+	}
+
+	return ret, errs
+}
+
+// PostRequestGroup behaves like agent.Post() but takes a group of URLs and performs the
+// requests in parallel. The number of simultaneous requests is controlled by
+// options.MaxParallel.
+//
+// The list of URLs and postData byte arrays are expected to be of equal length.
+// If postData has less elements than the url list, those urls without a corresponding
+// postData array will return an error.
+func (a *Agent) PostRequestGroup(urls []string, postData [][]byte) ([]*http.Response, []error) {
```

Comment:
I'm thinking out loud, but why is this not a map instead of two slices? I believe it might make the implementation easier, but also make it easier and less error prone to use this function.

Reply:
Mostly to keep the function signature in sync with the GET and the Group variants. It is a matrix of function interfaces which (at least in my mind) is easy to reason when you need to pick which one to use. It's similar to the way the regexp package defines its methods:

```
There are 16 methods of Regexp that match a regular expression and identify
the matched text. Their names are matched by this regular expression:

Find(All)?(String)?(Submatch)?(Index)?
```

If we turn it into a map, how would you define the response? map of slices?

Reply:
> If we turn it into a map, how would you define the response? map of slices?

Good question 🤔 I think a map of responses might make sense, i.e. `map[string]*http.Response`, where the key is the URL that was used to make the request (i.e. both maps (the argument and the results maps) would use the same key set).

Reply:
And the errors? Another `map[url]error`?

We can also create *Map variants, eg:

```golang
PostMap(map[string][]byte) 
```
 if those make more sense but I would prefer not to lose the plural variations:

```golang

// Singular
Post(url string, postData[]byte) ([]byte, error) 

// Plural variant
PostGroup(urls []string, postData [][]byte) ([][]byte, []error)
```

.. especially since it would take a different argument than what we have for GetGroup:

```golang
GetGroup(urls []string)
```

Reply:
Yeah, it gets complicated... Maybe `map[string]Response` would do better, where response is something like
```go
type Response struct {
        HttpResponse *http.Response
        err error
}
```

But I'm not sure this looks nice 😅

Reply:
OK, but can we do the maps and different responses in other variants of these functions?

Reply:
+1, let's leave it as it is and we'll figure out if we need to change anything at all.